explorer/service/src/main/java/org/radixdlt/explorer/metrics/MetricsProvider.java

```diff
@@ -178,25 +170,62 @@ private void calculateMetrics(Map<String, SystemInfo> info) {
         long progress = Math.round(aggregatedProgress / nodeCount);
 
         synchronized (calculationLock) {
-            calculationIterations++;
             long seconds = (System.currentTimeMillis() - testState.getStartTimestamp()) / 1000 + 1;
             assert seconds > 0;
             averageTps = Math.round(progress / seconds);
             peakTps = Math.max(peakTps, tps);
             calculatedMetrics = new Metrics(tps, progress, averageTps, peakTps);
+            dumpCurrentMetrics();
         }
 
+        subject.onNext(calculatedMetrics);
+    }
+
+    /**
+     * Resets the metrics dump file to only contain a single header line,
+     * if a path to it has been set.
+     */
+    private void resetDumpFile() {
         if (metricsDumpPath != null) {
             try {
-                String line = calculatedMetrics.toString();
+                byte[] data = Metrics.DATA_HEADLINE.getBytes(UTF_8);
+                Files.write(metricsDumpPath, data, CREATE, WRITE);
+            } catch (Exception e) {
+                LOGGER.info("Couldn't reset metrics dump file: " + metricsDumpPath, e);
+            }
+        }
+    }
+
+    /**
+     * Dumps the current metrics to a file, if a path to it has been set.
+     */
+    private void dumpCurrentMetrics() {
+        if (metricsDumpPath != null) {
+            String line = calculatedMetrics.toString();
+            try {
                 byte[] data = line.getBytes(UTF_8);
                 Files.write(metricsDumpPath, data, CREATE, WRITE, APPEND);
             } catch (Exception e) {
-                LOGGER.info("Couldn't dump metrics to file: " + metricsDumpPath, e);
+                LOGGER.info("Couldn't dump metrics to file: " + line, e);
             }
         }
+    }
 
-        subject.onNext(calculatedMetrics);
+    /**
+     * Restores the current metrics from the last dumped data, if a path
+     * to the dump file has been set.
+     */
+    private void restoreMetrics() {
+        if (metricsDumpPath != null) {
+            try {
+                List<String> lines = Files.readAllLines(metricsDumpPath, UTF_8);
```

Comment:
Only a good idea if you know for sure that the max file size **always** fits nicely in memory.

Reply:
Good point. I don't expect the file to grow "in absurdum" though as it's reset every time a test run starts.

Reply:
I have fixed this now (using a `RandomAccessFile` and reading from the end instead).